### PR TITLE
Track git ref sha in k8s CRDs by annotations

### PR DESCRIPTION
**Pull request by x10an14**

### CODE REVIEW

**Comment by x10an14** [comment]
Not yet ready - thought I'd made this PR as a draft. Apologies.

**Comment by tronghn** [comment]
psst: you can [convert a PR to a draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft)

File: actions/deploy/entrypoint.sh [hunk 1]
```diff
@@ -19,16 +19,20 @@ if [ -z "$WAIT" ]; then
...
+if [ -z "$GITHUB_SHA" ]; then
+    export GIT_REF_SHA="$GITHUB_SHA"
+fi
+
 # Inject "image" as a template variable to a new copy of the vars file.
 # If the file doesn't exist, it is created. The original file is left untouched.
 if [ ! -z "$IMAGE" ]; then
     export VARS_ORIGINAL="$VARS"
     export VARS=`mktemp`
+    cat "$VARS_ORIGINAL" > "$VARS"
```
Comment (ambientsound) [inline actions/deploy/entrypoint.sh:31]: What is the purpose?

File: pkg/deployd/deployd/deployd.go [hunk 2]
```diff
@@ -53,13 +53,15 @@ func jsonToResources(json []json.RawMessage) ([]unstructured.Unstructured, error
 }
 
 // Annotate a resource with the deployment correlation ID.
-func addCorrelationID(resource *unstructured.Unstructured, correlationID string) {
-	anno := resource.GetAnnotations()
-	if anno == nil {
-		anno = make(map[string]string)
+func AddAnnotations(resource *unstructured.Unstructured, annotations *map[string]string) {
```
Comment (ambientsound) [inline pkg/deployd/deployd/deployd.go:56]: - This function does not need to be public/exported (capital first letter in Go makes a function public)
- Maps are always passed by reference, no pointer needed

File: pkg/deployd/deployd/deployd.go [hunk 3]
```diff
@@ -138,7 +140,13 @@ func Run(logger *log.Entry, req *pb.DeploymentRequest, cfg config.Config, kube k
 	errors := make(chan error, len(resources))
 
 	for index, resource := range resources {
-		addCorrelationID(&resource, req.GetDeliveryID())
+		AddAnnotations(
+			&resource,
+			&(map[string]string{
+				kubeclient.CorrelationIDAnnotation: req.GetDeliveryID(),
+				kubeclient.GitRefShaAnnotation:     req.GetGitRefSha(),
```
Comment (ambientsound) [inline pkg/deployd/deployd/deployd.go:147]: Perhaps also add the `GitRef` annotation can be useful here?

File: pkg/hookd/database/deployment.go [hunk 4]
```diff
@@ -11,6 +11,7 @@ type Deployment struct {
 	Created          time.Time
 	GitHubID         *int
 	GitHubRepository *string
+	GitRefSha        *string
```
Comment (ambientsound) [inline pkg/hookd/database/deployment.go:14]: Not needed unless this value goes into the database, in which case you'll have to add a database migration.

File: actions/deploy/entrypoint.sh (hunk 1 above)
Comment (x10an14) [reply actions/deploy/entrypoint.sh:31]: Reduce cyclomatic complexity https://en.wikipedia.org/wiki/Cyclomatic_complexity#Implications_for_software_testing
As mentioned in the commit message =)

File: pkg/deployd/deployd/deployd.go (hunk 2 above)
Comment (x10an14) [reply pkg/deployd/deployd/deployd.go:56]: Nice! Nyttig å vite! =)

Comment (x10an14) [reply pkg/deployd/deployd/deployd.go:56]: Done!

File: pkg/deployd/deployd/deployd.go (hunk 3 above)
Comment (x10an14) [reply pkg/deployd/deployd/deployd.go:147]: Done!

**Comment by x10an14** [comment]
> psst: you can [convert a PR to a draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft)

Hehehe, thanks @tronghn. I always forget where to check. I find it so anonymous/hidden =P
I prefer the `WIP:` prefix used by GitLab. It also helps that the "button" (read: link) to toggle is right beside PR title, and not to the right within a more cluttered menu.

**Pull request closed without merging**